## Multiple Linear Regression and Adjusted R-squared

### Import relevant Libraries

In [30]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
import seaborn
seaborn.set()

from sklearn.linear_model import LinearRegression

### Load the Data

In [17]:
data = pd.read_csv('1.02. Multiple linear regression GPA.csv')

In [18]:
data

SAT   GPA  Rand 1,2,3
0   1714  2.40           1
1   1664  2.52           3
2   1760  2.54           3
3   1685  2.74           3
4   1693  2.83           2
..   ...   ...         ...
79  1936  3.71           3
80  1810  3.71           1
81  1987  3.73           3
82  1962  3.76           1
83  2050  3.81           2

[84 rows x 3 columns]

In [19]:
data.describe()

SAT        GPA  Rand 1,2,3
count    84.000000  84.000000   84.000000
mean   1845.273810   3.330238    2.059524
std     104.530661   0.271617    0.855192
min    1634.000000   2.400000    1.000000
25%    1772.000000   3.190000    1.000000
50%    1846.000000   3.380000    2.000000
75%    1934.000000   3.502500    3.000000
max    2050.000000   3.810000    3.000000

### First multiple regression

In [20]:
# GPA = b0 + b1(SAT) + b2(Rand 1,2,3)
y = data['GPA']
x1 = data[['SAT','Rand 1,2,3']]

In [21]:
x = sm.add_constant(x1)
results = sm.OLS(y,x).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    GPA   R-squared:                       0.407
Model:                            OLS   Adj. R-squared:                  0.392
Method:                 Least Squares   F-statistic:                     27.76
Date:                Mon, 07 Mar 2022   Prob (F-statistic):           6.58e-10
Time:                        17:37:34   Log-Likelihood:                 12.720
No. Observations:                  84   AIC:                            -19.44
Df Residuals:                      81   BIC:                            -12.15
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.2960      0.417      0.710      0.480      -0.533       1.125
SAT            0.0017      0.000      7.432      0.000       0.001       0.002
Rand 1,2,3    -0.0083      0.027     -0.304      0.762      -0.062       0.046
==============================================================================
Omnibus:                       12.992   Durbin-Watson:                   0.948
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               16.364
Skew:                          -0.731   Prob(JB):                     0.000280
Kurtosis:                       4.594   Cond. No.                     3.33e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.33e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

<br />

# Standardization

In [22]:
from sklearn.preprocessing import StandardScaler

In [23]:
scaler = StandardScaler()

In [24]:
scaler.fit(x1)

StandardScaler()

In [25]:
x1_scaled = scaler.transform(x1)

In [26]:
x1_scaled

array([[-1.26338288, -1.24637147],
       [-1.74458431,  1.10632974],
       [-0.82067757,  1.10632974],
       [-1.54247971,  1.10632974],
       [-1.46548748, -0.07002087],
       [-1.68684014, -1.24637147],
       [-0.78218146, -0.07002087],
       [-0.78218146, -1.24637147],
       [-0.51270866, -0.07002087],
       [ 0.04548499,  1.10632974],
       [-1.06127829,  1.10632974],
       [-0.67631715, -0.07002087],
       [-1.06127829, -1.24637147],
       [-1.28263094,  1.10632974],
       [-0.6955652 , -0.07002087],
       [ 0.25721362, -0.07002087],
       [-0.86879772,  1.10632974],
       [-1.64834403, -0.07002087],
       [-0.03150724,  1.10632974],
       [-0.57045283,  1.10632974],
       [-0.81105355,  1.10632974],
       [-1.18639066,  1.10632974],
       [-1.75420834,  1.10632974],
       [-1.52323165, -1.24637147],
       [ 1.23886453, -1.24637147],
       [-0.18549169, -1.24637147],
       [-0.5608288 , -1.24637147],
       [-0.23361183,  1.10632974],
       [ 1.68156984,

### regression with scaled features

In [31]:
reg = LinearRegression()
reg.fit(x1_scaled,y)

LinearRegression()

In [32]:
reg.coef_

array([ 0.17181389, -0.00703007])

In [33]:
reg.intercept_

3.330238095238095

## Summary Table

In [37]:
reg_summary = pd.DataFrame([['Bias'],['SAT'],['Rand 1,2,3']], columns=['Features'])
reg_summary['weights'] = reg.intercept_, reg.coef_[0], reg.coef_[1]

In [38]:
reg_summary

Features   weights
0        Bias  3.330238
1         SAT  0.171814
2  Rand 1,2,3 -0.007030

### Making predictions with standardized coefficients (weights)

In [39]:
new_data = pd.DataFrame(data=[[1700,2],[1800,1]], columns=['SAT','Rand 1,2,3'])
new_data

SAT  Rand 1,2,3
0  1700           2
1  1800           1

In [41]:
reg.predict(new_data)

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sklearn/base.py:443: UserWarning: X has feature names, but LinearRegression was fitted without feature names
  warnings.warn(


array([295.39979563, 312.58821497])

The error for teh aboe output occured because the algorithm was calculating based on standardized inputs. As GPA is never 295.4 or 312.6

In [42]:
new_data_scaled = scaler.transform(new_data)
new_data_scaled

array([[-1.39811928, -0.07002087],
       [-0.43571643, -1.24637147]])

In [43]:
reg.predict(new_data_scaled)

array([3.09051403, 3.26413803])